In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd

# data 路径
path_train = './data/train.csv'  
path_test = './data/test.csv'

# 设置列名
column_name = ['ID', 'Age', 'Gender', 'Edu', 'Query_list']

#  pandas读入训练数据和测试数据
df_train = pd.read_csv(path_train, sep = '###__###', header = None, encoding = 'utf-8')
df_train = df_train.dropna()

df_test = pd.read_csv(path_test, sep = '###__###', header = None, encoding = 'utf-8')
df_test = df_test.dropna()

# 给train数据加上列名
df_train.columns =  column_name

In [2]:
df_train.head()

ID  Age  Gender  Edu  \
0  22DD920316420BE2DF8D6EE651BA174B    1       1    4   
1  43CC3AF5A8D6430A3B572337A889AFE4    2       1    3   
2  E97654BFF5570E2CCD433EA6128EAC19    4       1    0   
3  6931EFC26D229CCFCEA125D3F3C21E57    4       2    3   
4  E780470C3BB0D340334BD08CDCC3C71A    2       2    4   

                                          Query_list  
0  柔和双沟\t女生\t中财网首页 财经\thttp://pan.baidu.com/s/1pl...  
1  广州厨宝烤箱\t世情薄,人情恶,雨送黄昏花易落,晓风干,泪痕\t厦门酒店用品批发市场\t我只...  
2  钻石之泪耳机\t盘锦到沈阳\t旅顺公交\t辽宁阜新车牌\tbaidu\tk715\tk716...  
3  最受欢迎狗狗排行榜\t舶怎么读\t场景描 写范例\t三维绘图软件\t枣和酸奶能一起吃吗\t好...  
4  干槽症能自愈吗\t太太万岁叶舒心去没去美国\t干槽症\t右眼皮下面一直跳是怎么回事\t麦当劳...

In [3]:
df_train.isnull().any()

ID            False
Age           False
Gender        False
Edu           False
Query_list    False
dtype: bool

In [4]:
df_train.dtypes

ID            object
Age            int64
Gender         int64
Edu            int64
Query_list    object
dtype: object

In [5]:
df_train.describe()

Age         Gender            Edu
count  100000.000000  100000.000000  100000.000000
mean        2.082310       1.387140       3.903330
std         1.184276       0.529495       1.522205
min         0.000000       0.000000       0.000000
25%         1.000000       1.000000       3.000000
50%         2.000000       1.000000       4.000000
75%         3.000000       2.000000       5.000000
max         6.000000       2.000000       6.000000

In [6]:
df_train.loc[32620, 'Query_list']

'pan.baidu.com/s/1esytqsi\tpan.baidu.com/s/1mhbyjko\tpan.baidu.com/s/1jicqk7o\tpan.baidu.com/s/1qyvm8qw\tpan.baidu.com/s/1kuziaxd\tpan.baidu.com/s/1nv50c2h\tpan.baidu.com/s/1kvi5olp\tpan.baidu.com/s/1dfg7hjz\tpan.baidu.com/s/1nvrgidb\tpan.baidu.com/s/1nvvyabr\tpan.baidu.com/s/1ji9axqq\tpan.baidu.com/s/1qynd56w\tpan.baidu.com/s/1hsdbf6k\tpan.baidu.com/s/1o86kaii\tpan.baidu.com/s/1bojbiyr\tpan.baidu.com/s/1midafew\tpan.baidu.com/s/1pkys2kb\tpan.baidu.com/s/1nuxbyb3\tpan.baidu.com/s/1b0hvku\tpan.baidu.com/s/1kudx1yn\tpan.baidu.com/s/1slrlvyp\tpan.baidu.com/s/1miqayfu\tpan.baidu.com/s/1hrkvdus\tpan.baidu.com/s/1ku48yj9\tpan.baidu.com/s/1ddzwcox\tpan.baidu.com/s/1nuccqr3\tpan.baidu.com/s/1bpr74f1\tpan.baidu.com/s/1jijk3h0\tpan.baidu.com/s/1ktvsjmf\tpan.baidu.com/s/1pklw6ar\tpan.baidu.com/s/1bymckq\tpan.baidu.com/s/1slvmrtr\tpan.baidu.com/s/1mhumfh2\tpan.baidu.com/s/1qyal9bi\tpan.baidu.com/s/1midjjry\tpan.baidu.com/s/1kubuscj\tpan.baidu.com/s/1gfhmkhd\tpan.baidu.com/s/1pksniz5\tpan.baidu.com

##### 正则表达式去掉 http和www.等

In [7]:
import re

# 创建正则表达式函数
def remove_noise(document):
    noise_pattern = re.compile("|".join(["http\S+", ":http\S+", "ｗｗｗ.+?\t",'\d+\@\S+']))
    clean_text = re.sub(noise_pattern, "", document)
    return clean_text.strip()

# 使用创建的remove_noise函数,对Query_list列进行数据清洗
df_train['Query_list'] = df_train['Query_list'].map(remove_noise)

In [8]:
df_train.head()

ID  Age  Gender  Edu  \
0  22DD920316420BE2DF8D6EE651BA174B    1       1    4   
1  43CC3AF5A8D6430A3B572337A889AFE4    2       1    3   
2  E97654BFF5570E2CCD433EA6128EAC19    4       1    0   
3  6931EFC26D229CCFCEA125D3F3C21E57    4       2    3   
4  E780470C3BB0D340334BD08CDCC3C71A    2       2    4   

                                          Query_list  
0  柔和双沟\t女生\t中财网首页 财经\t\t周公解梦大全查询2345\t曹云金再讽郭德纲\t...  
1  广州厨宝烤箱\t世情薄,人情恶,雨送黄昏花易落,晓风干,泪痕\t厦门酒店用品批发市场\t我只...  
2  钻石之泪耳机\t盘锦到沈阳\t旅顺公交\t辽宁阜新车牌\tbaidu\tk715\tk716...  
3  最受欢迎狗狗排行榜\t舶怎么读\t场景描 写范例\t三维绘图软件\t枣和酸奶能一起吃吗\t好...  
4  干槽症能自愈吗\t太太万岁叶舒心去没去美国\t干槽症\t右眼皮下面一直跳是怎么回事\t麦当劳...

In [9]:
df_train.isnull().any()

ID            False
Age           False
Gender        False
Edu           False
Query_list    False
dtype: bool

##### 准备分词

In [10]:
import jieba

In [11]:
str_test = '财经\thttp://pan.baidu.com/s/1plpjtn9\t周公解梦大全查询2345\thttps://yunpan.cn/ocsqfgtfya2ewj'

In [12]:
print(jieba.lcut(str_test))

Building prefix dict from the default dictionary ...
Loading model from cache C:\Users\LCG\AppData\Local\Temp\jieba.cache
Loading model cost 0.482 seconds.
Prefix dict has been built successfully.


['财经', '\t', 'http', ':', '/', '/', 'pan', '.', 'baidu', '.', 'com', '/', 's', '/', '1plpjtn9', '\t', '周公', '解梦', '大全', '查询', '2345', '\t', 'https', ':', '/', '/', 'yunpan', '.', 'cn', '/', 'ocsqfgtfya2ewj']


##### 去停用词

In [13]:
path_stopword = './data/hit_stopwords.txt'

In [14]:
stopwords=pd.read_csv(path_stopword,index_col=False,quoting=3,sep="\t",names=['stopword'], encoding='utf-8')

In [15]:
stopwords=stopwords['stopword'].values

In [16]:
def preprocess_text(content_lines, sentences, category):
    for line in content_lines:
        try:
            segs=jieba.lcut(line)
            segs = list(filter(lambda x:len(x)>1, segs)) #没有解析出来的新闻过滤掉
            segs = list(filter(lambda x:x not in stopwords, segs)) #把停用词过滤掉
            sentences.append((" ".join(segs), category))
        except Exception as e:
            print(line)
            continue

In [17]:
def process_df_text(content_lines):       
    segs=jieba.lcut(content_lines)
    segs = list(filter(lambda x:len(x)>1, segs)) #没有解析出来的新闻过滤掉
    segs = list(filter(lambda x:x not in stopwords, segs)) #把停用词过滤掉
    return (" ".join(segs))

In [ ]:
df_train['Query_list'] = df_train['Query_list'].map(process_df_text)

##### 把处理后的Train数据写入文件

In [ ]:
file_proce_path = './data/Pro_Train_Data.csv'

In [ ]:
df_train.to_csv(file_proce_path,encoding='utf-8',index=False)

### 数据按照标签处理

#### 1、按Age处理

In [19]:
df_train.head()

ID  Age  Gender  Edu  \
0  22DD920316420BE2DF8D6EE651BA174B    1       1    4   
1  43CC3AF5A8D6430A3B572337A889AFE4    2       1    3   
2  E97654BFF5570E2CCD433EA6128EAC19    4       1    0   
3  6931EFC26D229CCFCEA125D3F3C21E57    4       2    3   
4  E780470C3BB0D340334BD08CDCC3C71A    2       2    4   

                                          Query_list  
0  柔和 双沟 女生 中财网 首页 财经 周公 解梦 大全 查询 2345 曹云金 再讽 郭德纲...  
1  广州 厨宝 烤箱 世情 人情 雨送 黄昏 花易落 风干 泪痕 厦门 酒店用品 批发市场 不想...  
2  钻石 之泪 耳机 盘锦 沈阳 旅顺 公交 辽宁 阜新 车牌 baidu k715 k716 ...  
3  受欢迎 狗狗 排行榜 场景 范例 三维 绘图 软件 酸奶 一起 壮观 衣服 网站 动漫 绘图...  
4  干槽症 自愈 太太 万岁 舒心 美国 干槽症 眼皮 下面 怎么回事 麦当劳 旋风 勺子 吉林...

In [20]:
df_train['Age'].value_counts()

1    38996
2    26744
3    18529
4    10654
5     2922
0     1666
6      489
Name: Age, dtype: int64

In [21]:
df_train_Age = df_train[['Age' , 'Query_list']]

In [22]:
df_train_Age.head()

Age                                         Query_list
0    1  柔和 双沟 女生 中财网 首页 财经 周公 解梦 大全 查询 2345 曹云金 再讽 郭德纲...
1    2  广州 厨宝 烤箱 世情 人情 雨送 黄昏 花易落 风干 泪痕 厦门 酒店用品 批发市场 不想...
2    4  钻石 之泪 耳机 盘锦 沈阳 旅顺 公交 辽宁 阜新 车牌 baidu k715 k716 ...
3    4  受欢迎 狗狗 排行榜 场景 范例 三维 绘图 软件 酸奶 一起 壮观 衣服 网站 动漫 绘图...
4    2  干槽症 自愈 太太 万岁 舒心 美国 干槽症 眼皮 下面 怎么回事 麦当劳 旋风 勺子 吉林...

In [23]:
df_train_Age['Age'].value_counts()

1    38996
2    26744
3    18529
4    10654
5     2922
0     1666
6      489
Name: Age, dtype: int64

##### 全部数据用来做训练

###### 首先切分数据

In [24]:
from sklearn.model_selection import train_test_split

X_Age = df_train_Age['Query_list'].values.tolist()
Y_Age = df_train_Age['Age'].values.tolist()

###### 使用词向量NB模型

In [25]:
from sklearn.feature_extraction.text import CountVectorizer

# 字符级装换为词向量
vec = CountVectorizer(
    ngram_range=(1,4) , 
    analyzer = 'word' ,    # 可调的参数
    max_features = 20000 ,  # 可调的参数
)

In [26]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import StratifiedKFold
import numpy as np

In [27]:
def stratifiedkfold_cv(x , y , vec_word , clf , shuffle = True , n_folds = 5):
    stratifiedk_fold = StratifiedKFold(n_splits = n_folds , shuffle = shuffle)
    history_score = []
    vec_word.fit(x)
    
    for train_index , test_index in stratifiedk_fold.split(x,y):
        X_train , X_test = x[train_index] , x[test_index]
        y_train , y_test = y[train_index] , y[test_index]
        clf.fit(vec_word.transform(X_train) , y_train)
        history_score.append(clf.score(vec_word.transform(X_test) , y_test))
        
    return history_score
    

In [ ]:
classifier = MultinomialNB()

history_score = stratifiedkfold_cv( np.array(X_Age) , np.array(Y_Age) , vec , classifier ) 

In [ ]:
print(history_score)

In [75]:
print('max_features = 20000')
print( history_score)

max_features = 20000
[0.50875, 0.50375, 0.50635, 0.50655, 0.5047]


In [ ]:
print('ngram_range = (1,4)')
print('max_features = 20000')
print('------------------------')
print( history_score)

#### 以上对Age标签的NB模型处理。当改变vec的参数时，可以提高性能：
- 当vec是默认参数时：（即analyzer = 'word' ,max_features = 4000 ,时）-->[0.4767, 0.47915, 0.4777, 0.4803, 0.4815]
- 当vec参数为       （analyzer = 'word' ,max_features = 20000时） -->[0.50875, 0.50375, 0.50635, 0.50655, 0.5047]
- 当vec参数为(ngram_range=(1,4) ,analyzer = 'word' , max_features = 20000时）-->    